### ISSUE TITLE
root objects (services) hang when used with Flask

### ISSUE THREAD

**Issue by minusplusminus** [tomerfiliba/rpyc#342]
Hi everyone,

I have some issues regarding threading combined with flask. 
Flask demands that everything is done within the context of Flask. To do that I must use:

[https://flask.palletsprojects.com/en/1.0.x/api/#flask.has_app_context](flask.has_app_context)

In my project I use 
[https://github.com/tomerfiliba/rpyc/tree/master/demos/chat](chatdemo)
as inspiration for the system

When I try to send using an exposed function, the application hangs for a short while and shows the above error. 

The data goes like this:
- client sends data from client to server
- server processes it, and sends it to component that needs it
this goes always correct. 
- client receives is from server. client executes it with `flask.has_app_context`. App hangs with message `error result expired`
Then I try to do it on a new thread, using `threading`. The other client receives the message, but the thread was somehow not closed. Gives following error:

```
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/tmp/pycharm_project_44/mobyWeb/utils.py", line 161, in msg_async
    get_controller_client().send(comp, msg)
  File "/tmp/pycharm_project_44/mobyWeb/utils.py", line 150, in get_controller_client
    callback_update=halclient_update_callback,
  File "/tmp/pycharm_project_44/mobylib/messaging/client.py", line 59, in connect
    self.requireState)
  File "venv/lib/python3.7/site-packages/rpyc/core/netref.py", line 242, in __call__
    return syncreq(_self, consts.HANDLE_CALL, args, kwargs)
  File "venv/lib/python3.7/site-packages/rpyc/core/netref.py", line 76, in syncreq
    return conn.sync_request(handler, proxy, *args)
  File "venv/lib/python3.7/site-packages/rpyc/core/protocol.py", line 455, in sync_request
    return self.async_request(handler, *args, timeout=timeout).value
  File "venv/lib/python3.7/site-packages/rpyc/core/async_.py", line 102, in value
    raise self._obj
_get_exception_class.<locals>.Derived: Set changed size during iteration

========= Remote Traceback (1) =========
Traceback (most recent call last):
  File "venv/lib/python3.7/site-packages/rpyc/core/protocol.py", line 322, in _dispatch_request
    res = self._HANDLERS[handler](self, *args)
  File "venv/lib/python3.7/site-packages/rpyc/core/protocol.py", line 577, in _handle_call
    return obj(*args, **dict(kwargs))
  File "controller/run.py", line 113, in exposed_connect
    callback_error)
  File "controller/ComponentToken.py", line 21, in __init__
    self.broadcast("* Hello %s *" % (self.name,))
  File "controller/ComponentToken.py", line 90, in broadcast
    for tok in tokens:
RuntimeError: Set changed size during iteration
```

##### Environment

- rpyc 4.02
- python 3.7
- operating system Linux

##### Minimal example

Server:

```python
# from threading import RLock
from threading import RLock

broadcast_lock = RLock()  # thread lock
tokens = set()


class ComponentToken(object):
    def __init__(self, HalToken, name, callback_send, callback_update, callback_state, callback_error):
        super()
        self.name = name
        self.stale = False
        self.rt = None
        self.HalToken = HalToken

        self.callback_send = callback_send
        self.callback_update = callback_update
        self.callback_state = callback_state
        self.callback_error = callback_error

        self.broadcast("* Hello %s *" % (self.name,))
        tokens.add(self)

    def exposed_send(self, component, message):
        if self.stale:
            raise ValueError("Component token is stale")

        self.broadcast_component(component, message, 'send')

    def exposed_update(self, component, message):
        print('exposed update',component,message)
        if self.stale:
            raise ValueError("Component token is stale")
        self.broadcast_component(component, message, 'update')

    def exposed_error(self, component, message):
        if self.stale:
            raise ValueError("Component token is stale")
        self.broadcast_component(component, message, 'error')

    def exposed_state(self, message):
        if self.stale:
            raise ValueError("Component token is stale")
        self.broadcast("[%s] %s" % (self.name, message))

    def exposed_logout(self):
        if self.stale:
            return
        self.stale = True
        self.callback_send = None
        tokens.discard(self)
        self.broadcast("* Component disconnected %s *" % (self.name,))

    def broadcast_component(self, component, message, type):
        global tokens

        stale = set()

        if component == 'hal':
            if type == 'send':
                self.HalToken.on_message(message, self)
            if type == 'update':
                self.HalToken.on_update(message, self)
            if type == 'state':
                self.HalToken.on_state(message, self)
            if type == 'error':
                self.HalToken.on_error(message, self)
            return

        # with broadcast_lock:
        for tok in tokens:
            try:
                if tok.name == component:
                    if type == 'send':
                        tok.callback_send(message)
                    if type == 'update':
                        tok.callback_update(message)
                    if type == 'state':
                        tok.callback_state(message)
                    if type == 'error':
                        tok.callback_error(message)
            except:
                stale.add(tok)
        tokens -= stale

    def broadcast(self, message):
        global tokens
        stale = set()
        # with broadcast_lock:
        for tok in tokens:
            try:
                tok.callback_state(message)
            except:
                stale.add(tok)
        tokens -= stale

class HalToken(object):
    @staticmethod
    def on_message(message, token):
        global unixclient

        if 'type' in message:
            try:
                if 'state' not in message:
                    print(message)
                if unixclient is not None:
                    unixclient.sendData(message)

            except IOError as e:
                print('[driver] error socket', e)
                token.broadcast_component('webserver', str(e), 'error')

        else:
            print('keys missing in message : ', message)

    @staticmethod
    def on_update(message, token):
        print(message)

    @staticmethod
    def on_state(message, token):
        print(message)

    @staticmethod
    def on_error(message, token):
        print('on_state', message)


#
# Main service that shares exposed def with slave services
#
class HALService(Service):
    def on_connect(self, conn):
        self.token = None

    def on_disconnect(self, conn):
        if self.token:
            if self.token.rt is not None:
                self.token.rt.stop()

            self.token.exposed_logout()

    def exposed_connect(self,
                        component_name=str,
                        callback_send=None,
                        callback_update=None,
                        callback_state=None,
                        callback_error=None,
                        requireState=None):

        if self.token and not self.token.stale:
            raise ValueError("already connected")
        if component_name in components:
            print('connecting :' + component_name)
            self.token = ComponentToken(HalToken,
                                        component_name,
                                        async_(callback_send),
                                        async_(callback_update),
                                        callback_state,
                                        callback_error)

            return self.token
        else:
            raise ValueError("wrong Component name ")


def socket_receive(json):
    print(json)


if __name__ == '__main__':
    t = None
    try:

        print('running HAL')

        t = ThreadedServer(HALService, port=19912)
        t.start()

    except KeyboardInterrupt:
        print('Interrupted')
    try:
        t.close()
       # unixclient.close()
        sys.exit(0)
    except SystemExit:
        os._exit(0)

```

Client:

```python
import json
import time

import rpyc

rpyc.core.protocol.DEFAULT_CONFIG['allow_pickle'] = True


class HalConnectionError(Exception):
    """Some connection error while talking to HAL."""
    pass


class Client(object):
    def __init__(self, requireState=None):
        self.conn = None
        self.user = None
        self.requireState = requireState

    def disconnect(self):
        print('on_close')
        self.bgsrv.stop()
        if self.conn:
            try:
                self.user.logout()
            except:
                pass

            time.sleep(2)
            try:
                self.bgsrv.stop()
                self.conn.close()

                time.sleep(2)
                self.user = None
                self.conn = None
            except:
                print('close')

    def on_close(self, widget):
        print('on_close')
        self.bgsrv.stop()
        self.disconnect()

    def connect(self, clientName, callback_message, callback_update=None, callback_state=None, callback_error=None):

        try:
            self.conn = rpyc.connect('localhost', 19912, keepalive=True)
        except ConnectionRefusedError as err:
            raise HalConnectionError(err)

        self.name = clientName
        self.user = self.conn.root.connect(clientName,
                                           self.on_message,
                                           self.on_update,
                                           self.on_state,
                                           self.on_error,
                                           self.requireState)

        self.bgsrv = rpyc.BgServingThread(self.conn)
        self.callback = callback_message
        self.callback_update = callback_update

    #
    # called by the reactor whenever the connection has something to send
    #
    def bg_server(self, source=None, cond=None):

        if self.conn:
            self.conn.poll_all()
            return True
        else:
            return False

    #
    # sends the current message
    #
    def send(self, component, msg):
        json_dict = {**{'client': self.name}, **msg}
        self.user.send(component, json.dumps(json_dict))


  #
    # called by the server, with the text
    #
    def on_message(self, text):
        if self.callback: self.callback(text)

    def on_update(self, text):
        if self.callback_update: self.callback_update(text)

   def get_controller_client():
    """Return a client for the HAL server."""
    if 'hal_client' not in flask.g:
        flask.g.hal_client = HalClient(requireState=False)
        flask.g.hal_client.connect(
            clientName='mobyWeb',
            callback_message=halclient_message_callback,
            callback_update=halclient_update_callback,
        )

    return flask.g.hal_client
    def halclient_update_callback(msg):
             app = web.flask_app
             with app.app_context():
                 # Hangs after this line
                 get_controller_client().send('component', {'key': 'value'})
```

**Comment by minusplusminus** [comment]
I think I fixed it. Have not done a stress test, but rpyc and flask don't hang anymore. 

Rpyc does not have the same context. So `flask.g.hal_client` does not exist.  stored the returned value and requested it.
```python

flask_app.hal_client = get_controller_client()
app = web.flask_app
        with app.app_context():
           app.hal_client.send('component', {'key': True})
```

**Comment by comrumino** [comment]
Things I noticed:
1. you are passing methods to the `exposed_connect` method which will create a netref within the server to them. The preferred way to do this would be to define a `rpyc.Service` for the client and pass it as a parameter to [rpyc.connect](https://rpyc.readthedocs.io/en/latest/api/utils_factory.html#rpyc.utils.factory.connect) (`rpyc.connect` is an alias to `rpyc.utils.factory.connect`).
2. the responsibilities of methods are not very clear and the passing of self leads me to believe there is room for some design improvements. Particularly, if within `on_message` there is an `IOError` and it seems like it recursion would continue until max stack depth is hit or the result might timeout before then---the behavior is unclear at best.
 
Sorry for the delay in response. Closing, but feel free to open another issue if you have more questions. Also, self contained examples that I can run locally help me provide a better explaination
